packages/w3p/src/providers/wrapped/evm-wallet-connect.ts

```
@@ -0,0 +1,316 @@
+import { WalletConnectModal } from '@walletconnect/modal'
+import UniversalProvider from '@walletconnect/universal-provider'
+import { providers, utils } from 'ethers'
+
+import {
+  CHAIN_TYPES,
+  PROVIDER_EVENT_BUS_EVENTS,
+  PROVIDER_EVENTS,
+  PROVIDERS,
+} from '@/enums'
+import {
+  getEthExplorerAddressUrl,
+  getEthExplorerTxUrl,
+  requestAddEthChain,
+  requestSwitchEthChain,
+} from '@/helpers'
+import { Provider } from '@/provider'
+import type {
+  Chain,
+  ChainId,
+  EthereumProvider,
+  EthTransactionResponse,
+  ProviderProxy,
+  RawProvider,
+  TransactionResponse,
+  TxRequestBody,
+  WalletConnectInitArgs,
+} from '@/types'
+
+import { ProviderEventBus } from '../wrapped/_event-bus'
+
+export class WalletConnectEvmProvider
+  extends ProviderEventBus
+  implements ProviderProxy
+{
+  #rawProvider: UniversalProvider
+  #ethProvider: providers.Web3Provider
+
+  #w3Modal: WalletConnectModal
+
+  #chainId?: ChainId
+  #address?: string
+
+  readonly #projectId: string
+
+  constructor(provider: RawProvider) {
+    super()
+
+    if (!(provider as WalletConnectInitArgs).projectId) {
+      throw new Error('projectId is required for WalletConnect provider')
+    }
+
+    const { projectId } = provider as WalletConnectInitArgs
+    this.#projectId = projectId
+    this.#rawProvider = {} as UniversalProvider
+    this.#ethProvider = {} as providers.Web3Provider
+    this.#w3Modal = new WalletConnectModal({
+      projectId: this.#projectId,
+    })
+  }
+
+  static get providerType(): PROVIDERS {
+    return PROVIDERS.WalletConnect
+  }
+
+  get rawProvider() {
```

1. add return type 
2. we should throw the error (like not initialized or sth like this) from this method in case if `this.#rawProvider` is empty object

---

done